patroni/log.py

```diff
@@ -535,15 +536,27 @@ def run(self) -> None:
                     prev_record, record = record, None
                 else:
                     if prev_record and prev_record.thread == record.thread:
-                        if not (record.msg.startswith('no action. ') or record.msg.startswith('PAUSE: no action')):
+                        if self._is_heartbeat_msg(record):
+                            config = self._config or {}
+                            deduplicate_heartbeat_logs = config.get('deduplicate_heartbeat_logs', False)
+                            if record.msg == prev_hb_msg and deduplicate_heartbeat_logs:
+                                record = None
+                            else:
+                                prev_hb_msg = record.msg
```

Comment:
Now it is much easier to understand. Thank you!
However, it would be also nice to add respective unit tests for it.

You can take as an example a test that covers `Lock owner` not being removed when it is not followed by heartbeat message: https://github.com/patroni/patroni/blob/b4eab4897161eb15b37c6995991d76bf56de9ff7/tests/test_log.py#L87-L94

Since it is multithreaded, it will not be easy to check whether expected log_handler.handle() calls are made, but it is not required. At least I want to see that all added lines are covered.